### PR TITLE
Feature/feature veterinary

### PR DESCRIPTION
**Pull request by xavier-campos-diaz**

### CODE REVIEW

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ app/src/main/java/org/pesmypetcare/mypetcare/features/pets/Event.java  11
         

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

**Review by enrichp — DISMISSED** [review]

**Review by enrichp — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ app/src/main/java/org/pesmypetcare/mypetcare/features/pets/Event.java  11
         

Complexity increasing per file
==============================
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

**Review by danicm47 — DISMISSED** [review]

**Review by enrichp — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

**Review by enrichp — DISMISSED** [review]

**Review by danicm47 — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsManagerService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    List<VetVisit> findVetVisitsByPet(User user, Pet pet) throws ExecutionException, InterruptedException;
+
+    /**
+     * Creates and adds a new vet visit to the given pet.
+     * @param user The owner of the pet
+     * @param pet The pet where the vet visits has to be added
+     * @param vetVisit The vet visit that has to be added to the pet
+     */
+    void createVetVisit(User user, Pet pet, VetVisit vetVisit) throws VetVisitAlreadyExistingException, ExecutionException, InterruptedException;
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsManagerService.java:31]: Codacy Issue found: [Line is longer than 120 characters (found 145).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

File: app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     * @return True if the meal was added successfully or false otherwise
+     */
+    public boolean isResult() {
+        return result;
+    }
+
+    /**
+     * Executes the transaction.
+     */
+    public void execute() throws VetVisitAlreadyExistingException, NotPetOwnerException, ExecutionException, InterruptedException {
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java:61]: Codacy Issue found: [Line is longer than 120 characters (found 131).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        String petName = pet.getName();
+        VetVisitData visitData = new VetVisitData(vetVisit.getReason(), vetVisit.getAddress());
+        org.pesmypetcare.usermanagerlib.datacontainers.VetVisit libraryVisit =
+            new org.pesmypetcare.usermanagerlib.datacontainers.VetVisit(vetVisit.getVisitDate().toString(), visitData);
+        ServiceLocator.getInstance().getPetManagerClient().deleteFieldCollectionElement(accessToken, owner, petName,
+            PetData.VET_VISITS, libraryVisit.getKey());
+    }
+
+    @Override
+    public void updateVetVisitKey(User user, Pet pet, String newDate, DateTime visitDate) throws ExecutionException, InterruptedException {
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java:60]: Codacy Issue found: [Line is longer than 120 characters (found 139).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    public List<VetVisit> findVetVisitsByPet(User user, Pet pet) throws ExecutionException, InterruptedException {
+        String accessToken = user.getToken();
+        String owner = user.getUsername();
+        String petName = pet.getName();
+        ArrayList<VetVisit> appVisits = new ArrayList<>();
+        List<org.pesmypetcare.usermanagerlib.datacontainers.VetVisit> libraryVisit =
+            ServiceLocator.getInstance().getPetCollectionsManagerClient().getAllVetVisits(accessToken, owner, petName);
+        for (org.pesmypetcare.usermanagerlib.datacontainers.VetVisit visit : libraryVisit) {
+            appVisits.add(new VetVisit(DateTime.Builder.buildFullString(visit.getKey()), visit.getBody().getAddress(),
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java:27]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            new org.pesmypetcare.usermanagerlib.datacontainers.VetVisit(newDate, vetVisitData);
+        ServiceLocator.getInstance().getPetManagerClient().deleteFieldCollectionElement(accessToken, owner, petName,
+            PetData.VET_VISITS, oldLibraryVisit.getKey());
+        ServiceLocator.getInstance().getPetManagerClient().addFieldCollectionElement(accessToken, owner, petName,
+            PetData.VET_VISITS, newLibraryVisit.getKey(), newLibraryVisit.getBodyAsMap());
+
+    }
+
+    @Override
+    public void updateVetVisitBody(User user, Pet pet, VetVisit vetVisit) throws ExecutionException, InterruptedException {
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java:78]: Codacy Issue found: [Line is longer than 120 characters (found 123).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsManagerService.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    void deleteVetVisit(User user, Pet pet, VetVisit vetVisit) throws ExecutionException, InterruptedException;
+
+    /**
+     * Updates the key of the visit date for the given visit date.
+     * @param user The owner of the pet
+     * @param pet The pet from which we want to update the visit
+     * @param newDate The new date to set to the visit
+     * @param visitDate The current date of the visit
+     */
+    void updateVetVisitKey(User user, Pet pet, String newDate, DateTime visitDate) throws ExecutionException, InterruptedException;
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsManagerService.java:48]: Codacy Issue found: [Line is longer than 120 characters (found 131).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            PetData.VET_VISITS, libraryVisit.getKey());
+    }
+
+    @Override
+    public void updateVetVisitKey(User user, Pet pet, String newDate, DateTime visitDate) throws ExecutionException, InterruptedException {
+        String accessToken = user.getToken();
+        String owner = user.getUsername();
+        String petName = pet.getName();
+        VetVisitData vetVisitData = ServiceLocator.getInstance().getPetCollectionsManagerClient().getVetVisit
+            (accessToken, owner, petName, visitDate.toString());
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java:65]: Codacy Issue found: ['(' should be on the previous line.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/VetVisitsControllersFactory.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/services/StubVetVisitsManagerService.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrDeleteVetVisit.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrNewVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/features/pets/VetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrObtainAllVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrDeleteVetVisit.java  3
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrObtainAllVetVisits.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TestTrUpdateVetVisit.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrUpdateVetVisit.java  4
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/vetvisits/TrNewVetVisit.java  3
         

Clones added
============
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetMealsFragment.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/fragments/infopet/InfoPetVetVisits.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/services/VetVisitsAdapter.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5493224&bid=18002073)

**Review by enrichp — APPROVED** [review]

**Review by danicm47 — APPROVED** [review]

**Pull request merged**